File: v4/rust/iprange-livedb/src/validation/range.rs
use crate::contract::{AddressFamily, ValueKind, MAX_TREE_LEVEL};
use crate::error::{Error, Result};
use crate::key::{IpKey, Ipv4Key, Ipv6Key};
use crate::mapping::{ByteRange, ByteSource, PageView};
use crate::range_tree::{self, Record as Range};
use crate::slotted_page::LayoutInspection;

use super::context::Context;
use super::membership_table::CountResult;
use super::page::{self, TreePageSpec};
use super::{ValidationObject, ValidationReason, ValidationSink};

pub(crate) fn validate<S: ValidationSink>(context: &mut Context<'_, S>) -> Result<()> {
    if context.meta.range_root == 0 {
        if context.meta.range_record_count != 0 {
            count_mismatch(context)?;
        }
        return Ok(());
    }

    let count = match context.meta.address_family {
        AddressFamily::Ipv4 => validate_family::<Ipv4Key, S>(context)?,
        AddressFamily::Ipv6 => validate_family::<Ipv6Key, S>(context)?,
    };
    if count != context.meta.range_record_count {
        count_mismatch(context)?;
    }
    Ok(())
}

fn validate_family<K: IpKey, S: ValidationSink>(context: &mut Context<'_, S>) -> Result<u64> {
    let mut state = RangeState::<K> {
        count: 0,
        previous: None,
    };
    let mut path = [0; MAX_TREE_LEVEL as usize + 1];
    let root = context.meta.range_root;
    validate_node(context, root, None, true, &mut path, 0, &mut state)?;
    Ok(state.count)
}

fn validate_node<K: IpKey, S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    expected_level: Option<u16>,
    root: bool,
    path: &mut [u32; MAX_TREE_LEVEL as usize + 1],
    depth: usize,
    state: &mut RangeState<K>,
) -> Result<Option<K>> {
    let Some(page) = read_range_page(context, page_number, path, depth)? else {
        state.previous = None;
        return Ok(None);
    };
    let Some((header, cells)) =
        validate_range_page::<K, S>(context, page_number, page, expected_level, root)?
    else {
        state.previous = None;
        return Ok(None);
    };
    if header.level == 0 {
        validate_leaf(context, page_number, cells, state)
    } else {
        validate_branch(context, page_number, cells, header, path, depth, state)
    }
}

fn read_range_page<'m, S: ValidationSink>(
    context: &mut Context<'m, S>,
    page_number: u32,
    path: &mut [u32; MAX_TREE_LEVEL as usize + 1],
    depth: usize,
) -> Result<Option<PageView<'m>>> {
    let Some(slot) = path.get_mut(depth) else {
        context.emit(
            ValidationReason::TreeLevelInvalid,
            ValidationObject::RangeTree,
            Some(page_number),
            None,
            None,
        )?;
        return Ok(None);
    };
    *slot = page_number;
    context.read_graph_page(page_number, ValidationObject::RangeTree, &path[..depth])
}

fn validate_range_page<'m, K: IpKey, S: ValidationSink>(
    context: &mut Context<'m, S>,
    page_number: u32,
    page: PageView<'m>,
    expected_level: Option<u16>,
    root: bool,
) -> Result<Option<(page::SlottedHeader, LayoutInspection<PageView<'m>>)>> {
    let Some(header) = page::slotted_header(
        context,
        page_number,
        page,
        ValidationObject::RangeTree,
        TreePageSpec {
            branch_type: range_tree::RANGE_BRANCH,
            leaf_type: range_tree::RANGE_LEAF,
            aux: K::FAMILY as u32,
            expected_level,
        },
    )?
    else {
        return Ok(None);
    };
    report_degenerate_root(context, page_number, header, root)?;
    let cell_len = range_cell_len::<K>(header.level);
    let Some(cells) = page::validate_fixed_cells(
        context,
        page_number,
        page,
        ValidationObject::RangeTree,
        header,
        cell_len,
    )?
    else {
        return Ok(None);
    };
    Ok(Some((header, cells)))
}

fn report_degenerate_root<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    header: page::SlottedHeader,
    root: bool,
) -> Result<()> {
    if root && header.level > 0 && header.item_count == 1 {
        context.emit(
            ValidationReason::TreeLevelInvalid,
            ValidationObject::RangeTree,
            Some(page_number),
            None,
            None,
        )?;
    }
    Ok(())
}

fn range_cell_len<K: IpKey>(level: u16) -> usize {
    if level == 0 {
        range_tree::record_size::<K>()
    } else {
        range_tree::branch_size::<K>()
    }
}

fn validate_leaf<K: IpKey, S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    cells: LayoutInspection<PageView<'_>>,
    state: &mut RangeState<K>,
) -> Result<Option<K>> {
    if context.meta.value_kind == ValueKind::Membership {
        validate_leaf_cells::<K, S, true>(context, page_number, cells, state)
    } else {
        validate_leaf_cells::<K, S, false>(context, page_number, cells, state)
    }
}

fn validate_leaf_cells<K: IpKey, S: ValidationSink, const MEMBERSHIP: bool>(
    context: &mut Context<'_, S>,
    page_number: u32,
    cells: LayoutInspection<PageView<'_>>,
    state: &mut RangeState<K>,
) -> Result<Option<K>> {
    let mut order = LeafOrder::default();
    for cell in cells.cells() {
        validate_leaf_cell::<K, S, _, MEMBERSHIP>(context, page_number, cell, &mut order, state)?;
    }
    Ok(order.first)
}

struct LeafOrder<K> {
    first: Option<K>,
    previous: Option<K>,
}

impl<K> Default for LeafOrder<K> {
    fn default() -> Self {
        Self {
            first: None,
            previous: None,
        }
    }
}

impl<K: IpKey> LeafOrder<K> {
    fn observe(&mut self, key: K) -> bool {
        self.first.get_or_insert(key);
        let invalid = self.previous.is_some_and(|previous| previous >= key);
        self.previous = Some(key);
        invalid
    }
}

#[allow(clippy::too_many_arguments)]
fn validate_leaf_cell<K: IpKey, S: ValidationSink, P: ByteSource, const MEMBERSHIP: bool>(
    context: &mut Context<'_, S>,
    page_number: u32,
    cell: ByteRange<P>,
    order: &mut LeafOrder<K>,
    state: &mut RangeState<K>,
) -> Result<()> {
    let current = read_range_cell::<K, _>(cell)?;
    if order.observe(current.from) {
        emit_range_finding(context, page_number, ValidationReason::TreeOrderInvalid)?;
    }
    // Every reachable page number and every per-page slot count fit in u32 and
    // u16 respectively, so the total range count cannot overflow u64.
    state.count += 1;
    if current.from > current.to {
        emit_range_finding(context, page_number, ValidationReason::RangeReversed)?;
        state.previous = None;
        return Ok(());
    }
    if MEMBERSHIP {
        if current.value == 0 {
            emit_range_finding(
                context,
                page_number,
                ValidationReason::MembershipBitmapInvalid,
            )?;
        } else {
            match context.count_membership_range(current.value) {
                CountResult::Full => emit_range_finding(
                    context,
                    page_number,
                    ValidationReason::MembershipRefcountInvalid,
                )?,
                CountResult::Cancelled => return Err(Error::Cancelled),
                CountResult::Unavailable => {
                    return Err(Error::Corrupt(
                        "membership validation has no membership table",
                    ));
                }
                CountResult::Inserted | CountResult::Existing => {}
            }
        }
    }
    if let Some(reason) = neighbor_problem(state.previous, current) {
        emit_range_finding(context, page_number, reason)?;
    }
    state.previous = Some(current);
    Ok(())
}

fn read_range_cell<K: IpKey, P: ByteSource>(cell: ByteRange<P>) -> Result<Range<K>> {
    range_tree::decode_fields(cell)
}

fn emit_range_finding<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    reason: ValidationReason,
) -> Result<()> {
    context.emit(
        reason,
        ValidationObject::RangeTree,
        Some(page_number),
        None,
        None,
    )
}

fn neighbor_problem<K: IpKey>(
    previous: Option<Range<K>>,
    current: Range<K>,
) -> Option<ValidationReason> {
    let previous = previous?;
    if current.from <= previous.to {
        Some(ValidationReason::RangeOverlap)
    } else if previous.to.checked_next() == Some(current.from) && previous.value == current.value {
        Some(ValidationReason::RangeNotCoalesced)
    } else {
        None
    }
}

fn validate_branch<K: IpKey, S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    cells: LayoutInspection<PageView<'_>>,
    header: page::SlottedHeader,
    path: &mut [u32; MAX_TREE_LEVEL as usize + 1],
    depth: usize,
    state: &mut RangeState<K>,
) -> Result<Option<K>> {
    let mut first = None;
    let mut previous = None;
    for cell in cells.cells() {
        let (key, child) = range_tree::decode_branch(cell)?;
        first.get_or_insert(key);
        if previous.is_some_and(|prior| prior >= key) {
            context.emit(
                ValidationReason::TreeOrderInvalid,
                ValidationObject::RangeTree,
                Some(page_number),
                None,
                None,
            )?;
        }
        previous = Some(key);
        let actual = validate_node(
            context,
            child,
            Some(header.level - 1),
            false,
            path,
            depth + 1,
            state,
        )?;
        if actual.is_some_and(|actual| actual != key) {
            context.emit(
                ValidationReason::TreeFenceInvalid,
                ValidationObject::RangeTree,
                Some(page_number),
                None,
                None,
            )?;
        }
    }
    Ok(first)
}

fn count_mismatch<S: ValidationSink>(context: &mut Context<'_, S>) -> Result<()> {
    context.emit(
        ValidationReason::RootCountInvalid,
        ValidationObject::RangeTree,
        None,
        None,
        None,
    )
}

struct RangeState<K> {
    count: u64,
    previous: Option<Range<K>>,
}

File: v4/go/internal/validation/range.go
package validation

// Range tree validation (Rust validation/range.rs): the family walk over
// the committed range tree with the record order, reversed, overlap, and
// coalescing findings, the value-kind reverse-count arms, and the root
// record-count proof. Every tree page is validated through the shared
// page authorities; every cell decodes through the format record codecs.

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/tree"
)

// validateRange runs the range tree validator (Rust range::validate): a
// zero root proves the record count is zero, any other root walks the
// selected family and the walk count must equal the declared record
// count.
func validateRange(ctx *context) error {
	if ctx.meta.RangeRoot == 0 {
		if ctx.meta.RangeRecordCount != 0 {
			return rangeCountMismatch(ctx)
		}
		return nil
	}
	var count uint64
	var err error
	switch ctx.meta.AddressFamily {
	case 4:
		count, err = validateRangeFamily4(ctx)
	case 6:
		count, err = validateRangeFamily6(ctx)
	default:
		// Unreachable through the bootstrap family check; kept as the
		// Rust enum exhaustiveness arm.
		return nil
	}
	if err != nil {
		return err
	}
	if count != ctx.meta.RangeRecordCount {
		return rangeCountMismatch(ctx)
	}
	return nil
}

func rangeCountMismatch(ctx *context) error {
	return ctx.emit(ReasonRootCountInvalid, ObjectRangeTree, nil, nil, nil)
}

// rangeState carries the walk count and the previous record of the
// neighbor checks across an ordered range walk (Rust RangeState).
type rangeState struct {
	count    uint64
	previous *format.RangeRecordV4
	prev6    *format.RangeRecordV6
	family   uint8
}

// validateRangeFamily4 walks the IPv4 range tree and returns the record
// count (Rust validate_family).
func validateRangeFamily4(ctx *context) (uint64, error) {
	state := rangeState{family: 4}
	var path [format.MaxTreeLevel + 1]uint32
	if _, _, err := walkRangeNode4(ctx, ctx.meta.RangeRoot, nil, true, &path, 0, &state); err != nil {
		return 0, err
	}
	return state.count, nil
}

// validateRangeFamily6 walks the IPv6 range tree and returns the record
// count.
func validateRangeFamily6(ctx *context) (uint64, error) {
	state := rangeState{family: 6}
	var path [format.MaxTreeLevel + 1]uint32
	if _, _, err := walkRangeNode6(ctx, ctx.meta.RangeRoot, nil, true, &path, 0, &state); err != nil {
		return 0, err
	}
	return state.count, nil
}

// rangePageHeader validates one range tree page and returns its header
// (Rust validate_range_page over page::slotted_header; the degenerate
// single-record root branch is the TreeLevelInvalid class).
func rangePageHeader(ctx *context, pageNumber uint32, page []byte, object ValidationObject, family uint8, expectedLevel *uint16, root bool) (*format.PageHeader, error) {
	header, err := treePageHeader(ctx, pageNumber, page, object, treePageSpec{
		branchType:    byte(format.PageTypeRangeBranch),
		leafType:      byte(format.PageTypeRangeLeaf),
		aux:           uint32(family),
		expectedLevel: expectedLevel,
	})
	if err != nil || header == nil {
		return nil, err
	}
	if err := validateRootShape(ctx, pageNumber, object, root, header); err != nil {
		return nil, err
	}
	return header, nil
}

func walkRangeNode4(ctx *context, pageNumber uint32, expectedLevel *uint16, root bool, path *[format.MaxTreeLevel + 1]uint32, depth int, state *rangeState) (tree.Key, bool, error) {
	page, err := readTreeNodePage(ctx, pageNumber, ObjectRangeTree, path, depth)
	if err != nil || page == nil {
		return tree.Key{}, false, err
	}
	header, err := rangePageHeader(ctx, pageNumber, page, ObjectRangeTree, 4, expectedLevel, root)
	if err != nil || header == nil {
		return tree.Key{}, false, err
	}
	cells, err := validateFixedCells(ctx, pageNumber, page, ObjectRangeTree, header, rangeCellLen4(header.Level))
	if err != nil || cells == nil {
		return tree.Key{}, false, err
	}
	if header.Level == 0 {
		return walkRangeLeaf4(ctx, pageNumber, cells, state)
	}
	return walkRangeBranch4(ctx, pageNumber, cells, header, path, depth, state)
}

func walkRangeNode6(ctx *context, pageNumber uint32, expectedLevel *uint16, root bool, path *[format.MaxTreeLevel + 1]uint32, depth int, state *rangeState) (tree.Key, bool, error) {
	page, err := readTreeNodePage(ctx, pageNumber, ObjectRangeTree, path, depth)
	if err != nil || page == nil {
		return tree.Key{}, false, err
	}
	header, err := rangePageHeader(ctx, pageNumber, page, ObjectRangeTree, 6, expectedLevel, root)
	if err != nil || header == nil {
		return tree.Key{}, false, err
	}
	cells, err := validateFixedCells(ctx, pageNumber, page, ObjectRangeTree, header, rangeCellLen6(header.Level))
	if err != nil || cells == nil {
		return tree.Key{}, false, err
	}
	if header.Level == 0 {
		return walkRangeLeaf6(ctx, pageNumber, cells, state)
	}
	return walkRangeBranch6(ctx, pageNumber, cells, header, path, depth, state)
}

// rangeCellLen4/6 mirror range.rs range_cell_len: the record size at
// level zero, the branch entry size above.
func rangeCellLen4(level uint16) int {
	if level == 0 {
		return format.RangeRecordV4Size
	}
	return format.RangeEntryV4Size
}

func rangeCellLen6(level uint16) int {
	if level == 0 {
		return format.RangeRecordV6Size
	}
	return format.RangeEntryV6Size
}

// walkRangeLeaf4 visits every leaf cell in order: the per-leaf key
// order, one range count per record, the reversed/overlap/coalescing
// findings, and the value-kind reverse-count arms (Rust validate_leaf +
// validate_leaf_cells).
func walkRangeLeaf4(ctx *context, pageNumber uint32, cells *format.LayoutInspection, state *rangeState) (tree.Key, bool, error) {
	var order leafOrder
	iterator := cells.Cells()
	for {
		cell, ok := iterator.Next()
		if !ok {
			break
		}
		record, err := format.DecodeRangeFieldsV4(cell)
		if err != nil {
			// The fixed layout proved the exact record size; a decode
			// failure here is the Rust Corrupt propagation.
			return tree.Key{}, false, formatError(err)
		}
		if order.observe(tree.Key{Lo: uint64(record.From)}) {
			if err := emitRangeFinding(ctx, pageNumber, ReasonTreeOrderInvalid); err != nil {
				return tree.Key{}, false, err
			}
		}
		if state.count == ^uint64(0) {
			return tree.Key{}, false, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation range record count"}
		}
		state.count++
		if record.From > record.To {
			if err := emitRangeFinding(ctx, pageNumber, ReasonRangeReversed); err != nil {
				return tree.Key{}, false, err
			}
			state.previous = nil
			continue
		}
		if err := validateRangeValue(ctx, pageNumber, record.Value, state); err != nil {
			return tree.Key{}, false, err
		}
		if state.previous != nil {
			if reason := neighborProblem4(state.previous, record); reason != 0 {
				if err := emitRangeFinding(ctx, pageNumber, reason); err != nil {
					return tree.Key{}, false, err
				}
			}
		}
		copyRecord := record
		state.previous = &copyRecord
	}
	return order.first, order.hasFirst, nil
}

// walkRangeLeaf6 is the IPv6 form of walkRangeLeaf4.
func walkRangeLeaf6(ctx *context, pageNumber uint32, cells *format.LayoutInspection, state *rangeState) (tree.Key, bool, error) {
	var order leafOrder
	iterator := cells.Cells()
	for {
		cell, ok := iterator.Next()
		if !ok {
			break
		}
		record, err := format.DecodeRangeFieldsV6(cell)
		if err != nil {
			return tree.Key{}, false, formatError(err)
		}
		if order.observe(tree.Key{Hi: record.FromHi, Lo: record.FromLo}) {
			if err := emitRangeFinding(ctx, pageNumber, ReasonTreeOrderInvalid); err != nil {
				return tree.Key{}, false, err
			}
		}
		if state.count == ^uint64(0) {
			return tree.Key{}, false, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation range record count"}
		}
		state.count++
		if record.FromHi > record.ToHi || (record.FromHi == record.ToHi && record.FromLo > record.ToLo) {
			if err := emitRangeFinding(ctx, pageNumber, ReasonRangeReversed); err != nil {
				return tree.Key{}, false, err
			}
			state.previous = nil
			state.prev6 = nil
			continue
		}
		if err := validateRangeValue(ctx, pageNumber, record.Value, state); err != nil {
			return tree.Key{}, false, err
		}
		if state.prev6 != nil {
			if reason := neighborProblem6(state.prev6, record); reason != 0 {
				if err := emitRangeFinding(ctx, pageNumber, reason); err != nil {
					return tree.Key{}, false, err
				}
			}
		}
		copyRecord := record
		state.prev6 = &copyRecord
		state.previous = nil
	}
	return order.first, order.hasFirst, nil
}

// leafOrder is the per-leaf first/previous key cursor (Rust LeafOrder).
type leafOrder struct {
	first       tree.Key
	hasFirst    bool
	previous    tree.Key
	hasPrevious bool
}

// observe records one key and reports a non-increasing key.
func (o *leafOrder) observe(key tree.Key) bool {
	if !o.hasFirst {
		o.first = key
		o.hasFirst = true
	}
	invalid := o.hasPrevious && !o.previous.Less(key)
	o.previous = key
	o.hasPrevious = true
	return invalid
}

// validateRangeValue runs the value-kind arm of one range record (Rust
// validate_leaf_cell KIND arms: Direct has none, Membership counts the
// value in the membership table, Structured in the structure table; a
// zero value is its missing class).
func validateRangeValue(ctx *context, pageNumber uint32, value uint32, state *rangeState) error {
	switch ctx.meta.ValueKind {
	case format.ValueKindMembership:
		if value == 0 {
			return emitRangeFinding(ctx, pageNumber, ReasonMembershipBitmapInvalid)
		}
		switch ctx.countMembershipOwner(value) {
		case CountFull:
			return emitRangeFinding(ctx, pageNumber, ReasonMembershipRefcountInvalid)
		case CountCancelled:
			return &format.Error{Code: format.CodeCancelled, Detail: "validation cancelled"}
		case CountUnavailable:
			return &format.Error{Code: format.CodeFormatInvalid, Detail: "membership validation has no membership table"}
		}
	case format.ValueKindStructured:
		if value == 0 {
			return emitRangeFinding(ctx, pageNumber, ReasonStructureMissing)
		}
		switch ctx.countStructureRange(value) {
		case CountFull:
			return emitRangeFinding(ctx, pageNumber, ReasonStructureRefcountInvalid)
		case CountCancelled:
			return &format.Error{Code: format.CodeCancelled, Detail: "validation cancelled"}
		case CountUnavailable:
			return &format.Error{Code: format.CodeFormatInvalid, Detail: "structured validation has no structure table"}
		}
	}
	return nil
}

// neighborProblem4 mirrors Rust neighbor_problem over the previous
// record: an overlapping start is the overlap class, an exactly
// adjacent start with the same value is the not-coalesced class.
func neighborProblem4(previous *format.RangeRecordV4, current format.RangeRecordV4) ValidationReason {
	if current.From <= previous.To {
		return ReasonRangeOverlap
	}
	if current.From == previous.To+1 && previous.Value == current.Value {
		return ReasonRangeNotCoalesced
	}
	return 0
}

// neighborProblem6 is the IPv6 form of neighborProblem4.
func neighborProblem6(previous *format.RangeRecordV6, current format.RangeRecordV6) ValidationReason {
	if current.FromHi < previous.ToHi || (current.FromHi == previous.ToHi && current.FromLo <= previous.ToLo) {
		return ReasonRangeOverlap
	}
	adjacent := previous.ToLo == ^uint64(0)
	if adjacent {
		adjacent = previous.ToHi != ^uint64(0) && current.FromHi == previous.ToHi+1 && current.FromLo == 0
	} else {
		adjacent = current.FromHi == previous.ToHi && current.FromLo == previous.ToLo+1
	}
	if adjacent && previous.Value == current.Value {
		return ReasonRangeNotCoalesced
	}
	return 0
}

// walkRangeBranch4 visits every branch cell in order: the per-page key
// order, the child subtree at the expected level, and the child-first
// fence (Rust validate_branch).
func walkRangeBranch4(ctx *context, pageNumber uint32, cells *format.LayoutInspection, header *format.PageHeader, path *[format.MaxTreeLevel + 1]uint32, depth int, state *rangeState) (tree.Key, bool, error) {
	var keys branchKeys
	expected := header.Level - 1
	iterator := cells.Cells()
	for {
		cell, ok := iterator.Next()
		if !ok {
			break
		}
		first, child, err := format.DecodeRangeEntryFieldsV4(cell)
		if err != nil {
			return tree.Key{}, false, formatError(err)
		}
		key := tree.Key{Lo: uint64(first)}
		if err := recordBranchKey(ctx, pageNumber, ObjectRangeTree, key, &keys); err != nil {
			return tree.Key{}, false, err
		}
		actual, hasActual, err := walkRangeNode4(ctx, child, &expected, false, path, depth+1, state)
		if err != nil {
			return tree.Key{}, false, err
		}
		if err := validateFence(ctx, pageNumber, ObjectRangeTree, key, actual, hasActual); err != nil {
			return tree.Key{}, false, err
		}
	}
	return keys.first, keys.hasFirst, nil
}

// walkRangeBranch6 is the IPv6 form of walkRangeBranch4.
func walkRangeBranch6(ctx *context, pageNumber uint32, cells *format.LayoutInspection, header *format.PageHeader, path *[format.MaxTreeLevel + 1]uint32, depth int, state *rangeState) (tree.Key, bool, error) {
	var keys branchKeys
	expected := header.Level - 1
	iterator := cells.Cells()
	for {
		cell, ok := iterator.Next()
		if !ok {
			break
		}
		firstHi, firstLo, child, err := format.DecodeRangeEntryFieldsV6(cell)
		if err != nil {
			return tree.Key{}, false, formatError(err)
		}
		key := tree.Key{Hi: firstHi, Lo: firstLo}
		if err := recordBranchKey(ctx, pageNumber, ObjectRangeTree, key, &keys); err != nil {
			return tree.Key{}, false, err
		}
		actual, hasActual, err := walkRangeNode6(ctx, child, &expected, false, path, depth+1, state)
		if err != nil {
			return tree.Key{}, false, err
		}
		if err := validateFence(ctx, pageNumber, ObjectRangeTree, key, actual, hasActual); err != nil {
			return tree.Key{}, false, err
		}
	}
	return keys.first, keys.hasFirst, nil
}

// emitRangeFinding streams one range finding of the page (Rust
// emit_range_finding).
func emitRangeFinding(ctx *context, pageNumber uint32, reason ValidationReason) error {
	return ctx.emit(reason, ObjectRangeTree, &pageNumber, nil, nil)
}

// formatError maps a format codec error to the typed Corrupt class (the
// Rust decode propagation class; unreachable for fixed cells because
// the layout inspection proved the exact extents).
func formatError(err error) error {
	if err == nil {
		return nil
	}
	return &format.Error{Code: format.CodeFormatInvalid, Detail: err.Error()}
}
